src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs

```diff
@@ -277,113 +277,57 @@ private void BoundsCheck(in SequencePosition start, in SequencePosition end)
-            return true;
-        }
-
-        internal bool TryGetReadOnlyMemory(out ReadOnlyMemory<T> memory)
-        {
-            GetTypeAndIndices(Start.GetInteger(), End.GetInteger(), out SequenceType type, out int startIndex, out int endIndex);
-
-            object startObject = Start.GetObject();
-            object endObject = End.GetObject();
-
-            Debug.Assert(startObject != null);
-            Debug.Assert(endObject != null);
-
-            int length = endIndex - startIndex;
-            if (startObject != endObject)
-            {
-                // Can't get ReadOnlyMemory from multi-block segments
-                memory = default;
-                return false;
-            }
-            else if (type == SequenceType.Array)
-            {
-                Debug.Assert(startObject is T[]);
-
-                memory = new ReadOnlyMemory<T>(Unsafe.As<T[]>(startObject));
-            }
-            else if (type == SequenceType.MemoryManager)
-            {
-                Debug.Assert(startObject is MemoryManager<T>);
-
-                memory = Unsafe.As<MemoryManager<T>>(startObject).Memory;
-            }
-            else if (typeof(T) == typeof(char) && type == SequenceType.String)
-            {
-                Debug.Assert(startObject is string);
-
-                var text = Unsafe.As<string>(startObject);
-                memory = (ReadOnlyMemory<T>)(object)text.AsMemory();
-            }
-            else // ReadOnlySequenceSegment
-            {
-                Debug.Assert(startObject is ReadOnlySequenceSegment<T>);
-
-                memory = Unsafe.As<ReadOnlySequenceSegment<T>>(startObject).Memory;
-            }
+            Debug.Assert(_sequenceStart.GetObject() != null);
 
-            memory = memory.Slice(startIndex, length);
+            int startIndex = GetIndex(_sequenceStart);
+            segment = new ArraySegment<T>((T[])_sequenceStart.GetObject(), startIndex, GetIndex(_sequenceEnd) - startIndex);
             return true;
         }
 
         internal bool TryGetString(out string text, out int start, out int length)
         {
-            if (typeof(T) != typeof(char))
-            {
-                start = 0;
-                length = 0;
-                text = null;
-                return false;
-            }
-
-            GetTypeAndIndices(Start.GetInteger(), End.GetInteger(), out SequenceType type, out int startIndex, out int endIndex);
-
-            if (type != SequenceType.String)
+            if (typeof(T) != typeof(char) || GetSequenceType() != SequenceType.String)
             {
                 start = 0;
                 length = 0;
                 text = null;
                 return false;
             }
 
-            start = startIndex;
-            length = endIndex - startIndex;
-            text = (string)Start.GetObject();
+            Debug.Assert(_sequenceStart.GetObject() != null);
```

Comment:
Why do we need this debug assert?

If _sequenceStart.GetObject() is null. I think we should set `text` to empty string and return true.

What happens if we call ToString() on `default` ReadOnlySequence? I think we would throw null reference exception because `text.Substring(0, 0)` would throw.

Reply:
The question is what should return next methods for **default** sequence?
1.  TryGetReadOnlySequenceSegment now it return false and nulls.
2.  TryGetArray now it return false and default ArraySegment
3. TryGetReadOnlyMemory now it return true and default ReadOnlyMemory
4. TryGetString now it return false and null string. I think it is correct because default sequence contains null string inside not empty.

Reply:
>4. TryGetString now it return false and null string. I think it is correct because default sequence contains null string inside, not empty.

I think that might be OK, but we would also need to update ToString() to avoid throwing a null ref exception:
https://github.com/dotnet/corefx/blob/master/src/System.Memory/src/System/Buffers/ReadOnlySequence.cs#L313

cc @pakrym, @KrzysztofCwalina - we need to make sure we are doing the right thing, consistenty, for `default`/`null` ReadOnlySequence. Suggestions?

Reply:
I think these all are fine.
`TryGetReadOnlyMemory` is interesting but considering we treat `default(ReadOnlyMemory)` as valid usable value I would say it's correct.

Reply:
> I think these all are fine.

I created issue about this https://github.com/dotnet/corefx/issues/29134

Reply:
> I think that might be OK, but we would also need to update ToString() to avoid throwing a null ref exception:

ToString tests that this method return false and will not throw a null ref exception.